### PR TITLE
feat: Add an option to hide adjustment hints outside of `unsafe` blocks and functions

### PR DESCRIPTION
**Pull request by WaffleLapkin**
As the title suggests: this PR adds an option (namely `rust-analyzer.inlayHints.expressionAdjustmentHints.hideOutsideUnsafe`) that allows to hide adjustment hints outside of `unsafe` blocks and functions:

Requested by @BoxyUwU <3

### CODE REVIEW

File: crates/hir/src/lib.rs [hunk 1]
```diff
@@ -107,13 +107,16 @@ pub use {
...
         builtin_attr::AttributeTemplate,
+        expr::Expr,
         find_path::PrefixKind,
         import_map,
         nameres::ModuleSource,
         path::{ModPath, PathKind},
         type_ref::{Mutability, TypeRef},
         visibility::Visibility,
+        DefWithBodyId,
     },
```
Comment (WaffleLapkin) [inline crates/hir/src/lib.rs:120]: This list is preceded with "please don't add items here", which makes me unsure if it's okay.

File: crates/ide/src/inlay_hints/adjustment.rs [hunk 2]
```diff
@@ -110,6 +121,59 @@ pub(super) fn hints(
     Some(())
 }
 
+fn is_inside_unsafe(sema: &Semantics<'_, RootDatabase>, node: &SyntaxNode) -> bool {
```
Comment (WaffleLapkin) [inline crates/ide/src/inlay_hints/adjustment.rs:124]: This seems to be working fine, but I'm left wondering if there is a better way...

File: crates/hir/src/lib.rs (hunk 1 above)
Comment (flodiebold) [reply crates/hir/src/lib.rs:120]: Yeah, the intention is for HIR to provide a facade to the internal APIs, so these four types are ones we wouldn't want to expose here. It'd be better to provide `is_inside_unsafe` on `Semantics`.

Comment (WaffleLapkin) [reply crates/hir/src/lib.rs]: Okay, done 👍🏻

File: crates/hir/src/semantics.rs [hunk 3]
```diff
@@ -1458,6 +1470,59 @@ impl<'db> SemanticsImpl<'db> {
...
+            }
+            Either::Left(_) => None,
+            Either::Right(it) => self.to_def(it).map(<_>::into).map(DefWithBodyId::VariantId),
+        };
+        let Some(def) = def else { return false };
+        let enclosing_node = enclosing_item.as_ref().either(|i| i.syntax(), |v| v.syntax());
+
+        if ast::Fn::cast(enclosing_node.clone()).and_then(|f| f.unsafe_token()).is_some() {
+            return true;
+        }
```
Comment (Veykril) [inline crates/hir/src/semantics.rs:1502]:
```suggestion
		if matches!(&enclosing_item, Either::Left(ast::Item::Fn(it)) if f.unsafe_token().is_some()) {
            return true;
        }
        let enclosing_node = enclosing_item.as_ref().either(|i| i.syntax(), |v| v.syntax());
```

File: crates/hir/src/semantics.rs [hunk 4]
```diff
@@ -1458,6 +1470,59 @@ impl<'db> SemanticsImpl<'db> {
             .map(|ty| ty.original.is_packed(self.db))
             .unwrap_or(false)
     }
+
+    fn is_inside_unsafe(&self, node: &SyntaxNode) -> bool {
```
Comment (Veykril) [inline crates/hir/src/semantics.rs:1474]:
```suggestion
    fn is_inside_unsafe(&self, node: &ast::Expr)-> bool {
```
This handles expressions so it should only take such

**Comment by Veykril** [comment]
This might not play completely nice with macro expansions, but we'll see
@bors r+

**Pull request merged**